src/main/java/net/minecraftforge/client/renderer/transparency/OITLevelRenderer.java

```diff
@@ -0,0 +1,515 @@
+            renderCall.drawOIT();
+        }
+
+        profilerfiller.popPush("oit_rendering_compose");
+
+        transparentOITRenderTarget.unbindWrite();
+
+        transparentOITRenderTarget.blitToScreen(Minecraft.getInstance().getWindow().getWidth(), Minecraft.getInstance().getWindow().getHeight());
+
+        profilerfiller.pop();
+    }
+
+    /**
+     * Callback to set up rendering to the OIT render target.
+     *
+     * @param shaderInstance The shader instance to set up.
+     * @param shouldUpload Whether the uniform values should be uploaded.
+     */
+    private void setupTransparentRendering(ShaderInstance shaderInstance, boolean shouldUpload) {
+        if (shaderInstance.OIT_ENABLE != null) {
+            shaderInstance.OIT_ENABLE.set(1);
+            if (shouldUpload) {
+                shaderInstance.OIT_ENABLE.upload();
+            }
+        }
+        else {
+            return;
+        }
+
+        transparentOITRenderTarget.bindWrite(true);
+
+        RenderSystem.enableDepthTest();
+        glDepthMask(false);
+        glEnable(GL_BLEND);
+        glBlendFuncSeparate(GL_ONE, GL_ONE, GL_ZERO, GL_ONE_MINUS_SRC_ALPHA);
+    }
+
+    /**
+     * Callback to clean up rendering to the OIT render target.
+     *
+     * @param shaderInstance The shader instance to clean up.
+     */
+    private void cleanUpTransparentRendering(ShaderInstance shaderInstance) {
+        if (shaderInstance.OIT_ENABLE != null) {
+            shaderInstance.OIT_ENABLE.set(0);
+        }
+
+        glDepthMask(true);
+        RenderSystem.defaultBlendFunc();
+        glDisable(GL_BLEND);
+    }
+
+    /**
+     * Defines a render call that can be queued up.
+     */
+    private interface IRenderCall {
+        void drawDirect();
+
+        void drawOIT();
+    }
+
+    /**
+     * Defines a render call that can be queued up, based on a render type.
+     */
+    private interface IRenderTypeBasedRenderCall extends IRenderCall {
+        RenderType renderType();
+    }
+
+    /**
+     * Defines a render call that can be queued up, based on a particle render type.
+     */
+    private interface IParticleRenderCall extends IRenderCall {
+        ParticleRenderType particleRenderType();
+    }
+
+    /**
+     * A render call for a CPU rendered buffer.
+     * Captures all related information to render the buffer.
+     *
+     * @param renderType The render type to use.
+     * @param renderedBuffer The rendered buffer to render.
+     * @param modelViewMatrix The model view matrix to use.
+     * @param projectionMatrix The projection matrix to use.
+     */
+    private record RenderedBufferRenderCall(RenderType renderType, BufferBuilder.RenderedBuffer renderedBuffer, Matrix4f modelViewMatrix, Matrix4f projectionMatrix) implements IRenderTypeBasedRenderCall {
+        @Override
+        public void drawDirect()
+        {
+            renderType.doRender(renderedBuffer, (shader) -> {}, (shader) -> {});
```

Comment:
Given how both of these are NO-OP `ShaderInstance` consumers and `ChunkGeometryRenderCall` and `ParticleRenderCall` also both have NO-OP `ShaderInstance` consumers we should just create and store a single `private static final Consumer<ShaderInstance> NO_OP_CONSUMER = shader -> {};` in the `OITLevelRenderer` and then reference that in all these places so that we only allocate a single object for all these consumers